src/interrupts/mod.rs

```diff
@@ -0,0 +1,577 @@
+    // We only have IRQ0-IRQ97 => [7:0] is enough => mask = 0x000000FF
+    let ipsr = (ipsr & 0x000000FF) - 16;
+    match ISRS[ipsr as usize] {
+        Some(ref mut isr) => isr(),
+        None => default_handler(ipsr as u8),
+    }
+                 
+}
+
+#[no_mangle]
+#[used]
+#[allow(private_no_mangle_statics)]
+static INTERRUPTS: [unsafe extern "C" fn(); 98] = [dispatcher; 98];
+
+
+static mut ISRS: [Option<Box<FnMut()>>; 98] =
+    [None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+     None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+     None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+     None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+     None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+     None, None, None, None, None, None, None, None, None, None, None, None, None, None, None,
+     None, None, None, None, None, None, None, None];
+
+/// Default interrupt handler
+static mut DEFAULT_HANDLER: Option<Box<FnMut(u8)>> = None;
+
+// Unreachable at the moment (only when interrupt was enabled before InterruptTable got ownership...)
+fn default_handler(irq: u8) {
+    unsafe {
+        match DEFAULT_HANDLER {
+            Some(ref mut handler) => handler(irq),
+            None => panic!("No default handler"),
+        }
+    }
+}
+
+/// The error type that can occur when handling with interrupts.
+#[derive(Debug)]
+pub enum Error {
+    /// The error type which is returned when an interrupt is registered that is already being used.
+    InterruptAlreadyInUse(InterruptRequest),
+}
+
+/// The `InterruptHandle` is used to access and configure an active interrupt.
+pub struct InterruptHandle<T> {
+    _data_type: PhantomData<T>,
+    irq: InterruptRequest,
+}
+
+impl<T> InterruptHandle<T> {
+    fn new(irq: InterruptRequest) -> Self {
+        InterruptHandle { 
+            irq: irq,
+            _data_type: PhantomData,
+        }
+    }
+}
+
+/// The `InterruptTable` guarantees safe and 'free of data races' use of interrupts.
+///
+/// To ensure that no data races can occur, it uses the Send and Sync concurrency concept from Rust.
+/// The `InterruptTable` can only be used in the `code(&mut InterruptTable)` function passed to the `scope` function,
+/// to ensure that `InterruptTable.drop()` is called.
+///
+/// # Examples
+/// ```
+/// interrupts::scope(nvic, |irq| { hprintln!("Default handler: {}", irq) },
+///     use stm32f7::interrupts::interrupt_request::InterruptRequest::Tim7;
+///     use interrupts::Priority::P1;
+///     |interrupt_table| {
+///
+///         let _ = interrupt_table.register(Tim7, P1, || {
+///         hprintln!("Interrupt handler for Tim7");
+///     });
+/// });
+/// ```
+pub struct InterruptTable<'a> {
+    _lifetime: PhantomData<&'a ()>,
+    nvic: &'static mut Nvic,
+    data: [Option<* mut ()>; 98],
```

Comment:
I don't think this needs to be an `Option` anymore, since we always insert a value now.